### PR TITLE
(modelsRepo): Initial commit: Azure.Iot.ModelsRepository 

### PR DESCRIPTION
**Pull request by azabbasi**

### CODE REVIEW

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Release History
+
+## 1.0.0-preview.1 (Unreleased)
+
+### New features
+
+- Initial preview of Azure.IoT.ModelRepository SDK
```
Comment (digimaun) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/CHANGELOG.md:7]: Plural "Models"

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Azure.Iot.ModelsRepository.csproj [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    <TargetFrameworks>$(RequiredTargetFrameworks)</TargetFrameworks>
+
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+  </PropertyGroup>
+
+  <!-- Nuget properties -->
+  <PropertyGroup>
+    <PackageTags>IoT;ModelsRepository;Pnp;DigitalTwins$(PackageCommonTags)</PackageTags>
+    <Description>SDK for the Azure IoT Models Repository</Description>
+    <Version>1.0.0-preview.1</Version>
```
Comment (digimaun) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Azure.Iot.ModelsRepository.csproj:14]: Should this be `1.0.0-beta.1`?

Comment (azabbasi) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Azure.Iot.ModelsRepository.csproj:14]: I think we use preview for .NET packages.

Comment (azabbasi) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Azure.Iot.ModelsRepository.csproj:14]: I double checked, we used preview for ADT and appconfiguration used preview as well.

File: .github/CODEOWNERS [hunk 3]
```diff
@@ -261,6 +261,10 @@
 # ServiceLabel: %Digital Twins %Service Attention
 /sdk/digitaltwins/                              @drwill-ms @timtay-microsoft @abhipsaMisra @vinagesh @azabbasi @bikamani @barustum @jamdavi
 
+# PRLabel: %IoT Model Repository
```
Comment (digimaun) [inline .github/CODEOWNERS:264]: One more plural change and related directory name below

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# Release History
+
+## 1.0.0-preview.1 (Unreleased)
+
+### New features
+
+- Initial preview of Azure.IoT.ModelsRepository SDK
```
Comment (digimaun) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/CHANGELOG.md:7]: nit: for the actual namespace the 't' is lowercase as in Iot.

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/Program.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                    Console.WriteLine("[{0:HH:mm:ss:fff}][{1}] {2}", DateTimeOffset.Now, e.Level, message),
+                level: EventLevel.Verbose);
+
+            await ResolveExistingAsync();
+            await TryResolveButNotFoundAsync();
+        }
+
+        private static async Task ResolveExistingAsync()
+        {
+            var dtmi = "dtmi:com:example:TemperatureController;1";
```
Comment (drwill-ms) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/Program.cs:29]: Why aren't we using snippets?

Comment (azabbasi) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/Program.cs:29]: We will, we have a task to add readme.md file for samples

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/ModelsRepositoryClient.cs [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        /// An IDictionary containing the model definition(s) where the key is the dtmi
+        /// and the value is the raw model definition string.
+        /// </returns>
+        /// <exception cref="RequestFailedException">Thrown when a resolution failure occurs.</exception>
+        /// <param name="dtmi">A well-formed DTDL model Id. For example 'dtmi:com:example:Thermostat;1'.</param>
+        /// <param name="cancellationToken">The cancellationToken.</param>
+        [SuppressMessage(
+            "Usage",
+            "AZC0015:Unexpected client method return type.",
+            Justification = "<Pending>")]
```
Comment (drwill-ms) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/src/ModelsRepositoryClient.cs:74]: We need to fill in a justification

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/RepositoryHandler.cs [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        private readonly Uri _repositoryUri;
+        private readonly ModelsRepositoryClientOptions _clientOptions;
+
+        public RepositoryHandler(Uri repositoryUri, ClientDiagnostics clientDiagnostics, ModelsRepositoryClientOptions options)
+        {
+            Argument.AssertNotNull(options, nameof(options));
+
+            _clientOptions = options;
+            _clientDiagnostics = clientDiagnostics;
+            _modelFetcher = repositoryUri.Scheme == "file"
```
Comment (drwill-ms) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/src/RepositoryHandler.cs:29]: Should "file" be a constant somewhere?

Comment (azabbasi) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/src/RepositoryHandler.cs:29]: it's never reused, but sure, we can have it in the constants file.

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/tests/ClientTests.cs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+{
+    public class ClientTests : ModelsRepositoryTestBase
+    {
+        [Test]
+        public void CtorOverloads()
+        {
+            string remoteUriStr = "https://dtmi.com";
+            Uri remoteUri = new Uri(remoteUriStr);
+
+            ModelsRepositoryClientOptions options = new ModelsRepositoryClientOptions();
```
Comment (drwill-ms) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/tests/ClientTests.cs:19]: var

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/tests/ClientTests.cs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+            Assert.AreEqual(ModelsRepositoryClient.DefaultModelsRepository, new ModelsRepositoryClient().RepositoryUri);
```
Comment (drwill-ms) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/tests/ClientTests.cs:21]: Can we use FluentAssertions please?

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/tests/ModelsRepositoryRecordedTestBase.cs [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        protected ModelsRepositoryClient GetClient(ModelsRepositoryTestBase.ClientType clientType, ModelsRepositoryClientOptions options = default)
+        {
+            if (options == null)
+            {
+                options = new ModelsRepositoryClientOptions();
+            }
+
+            return
+                clientType == ModelsRepositoryTestBase.ClientType.Local
+                ? InstrumentClient(
```
Comment (drwill-ms) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/tests/ModelsRepositoryRecordedTestBase.cs:38]: tab in

**Review by drwill-ms — APPROVED** [review]

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/tests/ClientTests.cs (hunk 9 above)
Comment (azabbasi) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/tests/ClientTests.cs:21]: in progress

**Review by digimaun — APPROVED** [review]

**Pull request merged**